### PR TITLE
SWIFT-1391 Add MongoCollection async/await API

### PR DESCRIPTION
**Pull request by kmahar**
This is a large PR but most of the content is quite straightforward, just more async versions of existing API methods, so hopefully it is not too painful to review.

### CODE REVIEW

File: Sources/MongoSwift/AsyncAwait/MongoClient+AsyncAwait.swift [hunk 1]
```diff
@@ -48,8 +48,8 @@ extension MongoClient {
      * Run the `listDatabases` command.
      *
      * - Parameters:
-     *   - filter: Optional `Document` specifying a filter that the listed databases must pass. This filter can be based
-     *     on the "name", "sizeOnDisk", "empty", or "shards" fields of the output.
+     *   - filter: Optional `BSONDocument` specifying a filter that the listed databases must pass. This filter can be
```
Comment (kmahar) [inline Sources/MongoSwift/AsyncAwait/MongoClient+AsyncAwait.swift:51]: I ended up doing some general cleanup of docstrings throughout the new API methods during this. I noticed some remaining instances of `Document` in existing API methods as well which these were copied from but I am going to make a separate PR to fix those up.

File: Tests/MongoSwiftTests/AsyncAwaitTestUtils.swift [hunk 2]
```diff
@@ -0,0 +1,137 @@
+#if compiler(>=5.5) && canImport(_Concurrency) && os(Linux)
```
Comment (kmahar) [inline Tests/MongoSwiftTests/AsyncAwaitTestUtils.swift:1]: I was starting to add a lot of utility functions so I ended up breaking all this out into a separate file. however, there are no changes in lines 24 through 90 so you do not need to re-review those.

File: Tests/MongoSwiftTests/AsyncAwaitTestUtils.swift [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+import NIO
+import TestsCommon
+import XCTest
+
+/// Temporary utility function until XCTest supports `async` tests.
+func testAsync(_ block: @escaping () async throws -> Void) {
+    let group = DispatchGroup()
+    group.enter()
+    Task.detached {
+        do {
```
Comment (kmahar) [inline Tests/MongoSwiftTests/AsyncAwaitTestUtils.swift:14]: I realized I needed to add a do/catch here, because otherwise if the call to `block` threw an error the Task would finish without ever calling `group.leave()` and so the tests would hang indefinitely.

File: Tests/MongoSwiftTests/AsyncAwaitTestUtils.swift [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            throw TestError(message: " reply missing version string: \(reply)")
+        }
+        return try ServerVersion(versionString)
+    }
+
+    internal func serverParameters() async throws -> BSONDocument {
+        try await self.db("admin").runCommand(["getParameter": "*"])
+    }
+
+    internal func topologyType() async throws -> TestTopologyConfiguration {
```
Comment (kmahar) [inline Tests/MongoSwiftTests/AsyncAwaitTestUtils.swift:108]: was fun to get to try out some of the new concurrency features in these next few methods and implicitly confirm the driver can handle parallel usage across tasks! (explicit test of that to come soon in another PR)

File: Tests/MongoSwiftTests/AsyncAwaitTests.swift [hunk 5]
```diff
@@ -123,21 +44,74 @@ final class AsyncAwaitTests: MongoSwiftTestCase {
...
+                    expect(doc).to(equal(["_id": 1]))
+                }
             }
         }
     }
 
     func testMongoDatabase() throws {
         testAsync {
             try await self.withTestNamespace { _, db, _ in
+                try await db.drop()
```
Comment (kmahar) [inline Tests/MongoSwiftTests/AsyncAwaitTests.swift:91]: I think I didn't notice this issue with my last PR because when you run the whole suite the test DB ends up getting cleaned up somewhere else, but running just the async/await tests multiple times in a row leads to failure here since test.foo wasn't being cleaned up.

File: Sources/MongoSwift/AsyncAwait/MongoCollection+AsyncAwait.swift [hunk 6]
```diff
@@ -0,0 +1,48 @@
+#if compiler(>=5.5) && canImport(_Concurrency) && os(Linux)
+/// Extension to `MongoCollection` to support async/await APIs.
+extension MongoCollection {
+    /**
+     *   Drops this collection from its parent database.
```
Comment (patrickfreed) [inline Sources/MongoSwift/AsyncAwait/MongoCollection+AsyncAwait.swift:5]:
```suggestion
     * Drops this collection from its parent database.
```

File: Tests/MongoSwiftTests/AsyncAwaitTestUtils.swift [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        async let helloReply = try self.db("admin").runCommand(["hello": 1])
+        async let shards = try self.db("config").collection("shards").find().get().toArray().get()
+        return try await TestTopologyConfiguration(helloReply: helloReply, shards: shards)
+    }
+
+    internal func getUnmetRequirement(_ testRequirement: TestRequirement) async throws -> UnmetRequirement? {
+        async let topologyType = try self.topologyType()
```
Comment (patrickfreed) [inline Tests/MongoSwiftTests/AsyncAwaitTestUtils.swift:115]: just curious, is `async let` required to make these three evaluate simultaneously? Or would invoking them directly in the argument position of `getUnmetRequirements` behave the same way?

side note: GH doesn't seem able to highlight `async let` properly yet lol

**Review by patrickfreed — APPROVED** [review]
LGTM! (mod one space)

File: Tests/MongoSwiftTests/AsyncAwaitTestUtils.swift (hunk 7 above)
Comment (kmahar) [reply Tests/MongoSwiftTests/AsyncAwaitTestUtils.swift:115]: it does compile to one-line it like:
```swift
await testRequirement.getUnmetRequirement(givenCurrent: try self.serverVersion(), try self.topologyType(), try self.serverParameters())
```

but the behavior seems different - I roughly tested it by inserting print statements in each method before and after their `async` work like:
```swift
    internal func serverParameters() async throws -> BSONDocument {
        print("getting server parameters")
        let params = try await self.db("admin").runCommand(["getParameter": "*"])
        print("returning server parameters")
        return params
    }
```

as originally written, the interleaving of "getting x" and "returning x" varies on each run, for example one output is:
```
getting topology type
getting server version
getting server parameters
returning server version
returning topology type
returning server parameters
```

and another is
```
getting server version
getting topology type
getting server parameters
returning server parameters
returning server version
returning topology type
```

 which makes since given they are being run in parallel.

but written as you ask about in one line, the interleaving is always the same and suggests serial evaluation of each called function:
```
getting server version
returning server version
getting topology type
returning topology type
getting server parameters
returning server parameters
```

File: Sources/MongoSwift/AsyncAwait/MongoCollection+Indexes+AsyncAwait.swift [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+    /**
+     * Retrieves a list of names of the indexes currently on this collection.
+     *
+     * - Parameters:
+     *   - session: Optional `ClientSession` to use when executing this command.
+     *
+     * - Returns: A `MongoCursor` over the index names.
+     *
+     * - Throws: `MongoError.LogicError` if the provided session is inactive.
```
Comment (isabelatkinson) [inline Sources/MongoSwift/AsyncAwait/MongoCollection+Indexes+AsyncAwait.swift:201]: General question, do we call out all possible errors that could be thrown in documentation? This method and the one above could encounter server/network errors e.g., but I'm not sure how that maps to our error types/if that's something we include in docs

**Review by isabelatkinson — COMMENTED** [review]
lgtm, just one question

File: Sources/MongoSwift/AsyncAwait/MongoCollection+Indexes+AsyncAwait.swift (hunk 8 above)
Comment (kmahar) [reply Sources/MongoSwift/AsyncAwait/MongoCollection+Indexes+AsyncAwait.swift:201]: At some point, @patrickfreed went through the whole API and called out what we thought the most common errors would be. we generally haven't tried to list them exhaustively, I think the goal has generally been to focus on common pitfalls and errors that could be specific to the use of the particular API method.

there is an error handling guide in the docs that talks about some of the general common errors users could see on any method, like a `ConnectionError` or `ServerSelectionError`:
https://mongodb.github.io/mongo-swift-driver/docs/current/MongoSwift/error-handling-guide.html

Comment (isabelatkinson) [reply Sources/MongoSwift/AsyncAwait/MongoCollection+Indexes+AsyncAwait.swift:201]: cool, makes sense to focus on errors that are relevant to usage. thanks for the clarification!

**Review by isabelatkinson — APPROVED** [review]
